### ISSUE TITLE
Bug: [no-unnecessary-condition] False positive on dynamic access of generic value

### ISSUE THREAD

**Issue by justinsmid** [typescript-eslint/typescript-eslint#10600]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=KYDwDg9gTgLgBASwHY2FAZgQwMbDgMQQBtUoAZBAZxgElUBbSmpMAVxgB4AVAPjgG8AUHEQNKALjgBBKFEwBPbj2FxKwTFGwALSdSjIA5oIC%2BgwaEiw46VkmwwEEJNeKkK1OsEYAheQGlgeUolAAoVflEvSgAaVXVNLThjSUISNHdaMWY2Tl5olQBrQIlpWQUOIvkIdDg8wQBKSRk5RV4BFWwnajgiCAB3NABhTDUAZXjtOABeOI1tADoYCDJ%2BoZHgEPqzEShgGFYoZwQxefRXNBCQ46966eUREUrKecoIeg2Qytupvmv6AG1KgBdAD8iwgoxg%2BiQBk24JWAygwzUcOQ2CIrAAJsBKCFeojkcBxnMtPV6vkRFtjEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK%2BSpPRRE0aB2iRwYAL4gtQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
export interface FilterListItemsInput<T> {
  items: Array<T>
  search: string
}

export function filterListItemsByKeys<T>(
  { items, search }: FilterListItemsInput<T>,
  keys: Array<keyof T>,
): Array<T> {
  const lowerCaseSearch = search.toLowerCase()

  return items.filter((item) =>
    keys.some((key) => item[key]?.toString().toLowerCase().includes(lowerCaseSearch)),
  )
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-condition": "error",
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expect the code above not to error because `item[key]` **can** be nullish, but the ESLint rule says it can't.

### Actual Result

`ESLint: Unnecessary optional chain on a non-nullish value.(@typescript-eslint/ no-unnecessary-condition)`

### Additional Info

I found https://github.com/typescript-eslint/typescript-eslint/issues/9754 which seems similar, though I don't think it's quite the same.
